zcash_primitives/src/primitives.rs

```diff
@@ -313,4 +328,25 @@ impl<E: JubjubEngine> Note<E> {
         // commitment to the x-coordinate is an injective encoding.
         self.cm_full_point(params).to_xy().0
     }
+
+    pub fn rcm(&self) -> E::Fs {
+        match self.rseed {
+            Rseed::BeforeZip212(rcm) => rcm,
+            Rseed::AfterZip212(rseed) => E::Fs::to_uniform(prf_expand(&rseed, &[0x04]).as_bytes()),
+        }
+    }
+
+    pub fn generate_or_derive_esk<R: RngCore + CryptoRng>(&self, rng: &mut R) -> E::Fs {
+        match self.rseed {
+            Rseed::BeforeZip212(_) => {
+                // create random 64 byte buffer
+                let mut buffer = [0u8; 64];
+                &rng.fill_bytes(&mut buffer);
+
+                // reduce to uniform value
+                E::Fs::to_uniform(&buffer[..])
+            }
+            Rseed::AfterZip212(rseed) => E::Fs::to_uniform(prf_expand(&rseed, &[0x05]).as_bytes()),
```

Comment:
Extract this into a separate function that can be used elsewhere:
```rust
/// Returns the derived `esk` if this note was created after ZIP 212 activated.
pub fn derive_esk(&self) -> Option<E::Fs> {
    match self.rseed {
        Rseed::BeforeZip212(_) => None,
        Rseed::AfterZip212(rseed) => (E::Fs::to_uniform(prf_expand(&rseed, &[0x05]).as_bytes())),
    }
}
```
and then here:
```suggestion
            Rseed::AfterZip212(_) => self.derive_esk().unwrap(),
```